flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java

```diff
@@ -105,8 +105,10 @@ private Table convertConnectorTable(ConnectorCatalogTable<?, ?> table) {
 
 	private Table convertCatalogTable(ObjectPath tablePath, CatalogTable table) {
 		Optional<TableFactory> tableFactory = catalog.getTableFactory();
-		TableSource<Row> tableSource = tableFactory.map(tf -> ((TableSourceFactory) tf).createTableSource(tablePath, table))
-			.orElse(TableFactoryUtil.findAndCreateTableSource(table));
+		TableSource<Row> tableSource = tableFactory.map(tf -> ((TableSourceFactory) tf).createTableSource(tablePath, table)).get();
```

Comment:
by comparing the logic here and `HiveTableFactory.createTableSource()`, I found there are duplicated calls on `TableFactoryUtil.findAndCreateTableSource(table)`. That made me wonder if any table factory implementation shouldn't call the generic table discovery service (a.k.a `TableFactoryUtil.findAndCreateTableSource`), and should just create tables that is specific to that table factory itself. 

Thus, the `HiveTableFactory.createTableSource` should just be as the following:

```
@Override
	public TableSource<Row> createTableSource(ObjectPath tablePath, CatalogTable table) {
		...

		if (!isGeneric) {
			return createInputFormatTableSource(tablePath, table);
		} else {
			return null;
		}
	}
```

Reply:
Good point. However, that's not what I meant to do. HiveTableFactory is responsible to generate table source/sink for both generic and hive tables, so it needs to take care of both. In DatabaseCalciteSchema.convertCatalogTable(), it calls TableFactoryUtil.findAndCreateTableSource(table) only if the catalog doesn't return a table factory object. Thus, there is no duplication per se. However, I did realize that the logic here is confusion, especially with the java Optional object. I will refactor the code a bit to make it clearer.